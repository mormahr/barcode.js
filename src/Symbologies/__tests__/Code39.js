import {
	NARROW_BAR,
	NARROW_SPACE,
	WIDE_BAR,
	WIDE_SPACE
} from "../../Core/Characters"
import {
	encodeCode39,
	Mapping
} from "../Code39"

describe("encodeCode39", function () {
	describe("Delimiters", function () {
		it("starts the barcode with a start/stop code", function () {
			const code = encodeCode39("")
			expect(code.substr(0, 9)).toEqual(Mapping["*"])
		})

		it("ends the barcode with a start/stop code", function () {
			const code = encodeCode39("")
			expect(code.substr(-9)).toEqual(Mapping["*"])
		})
	})

	it("separates codes with a narrow space", function () {
		const code = encodeCode39("")
		expect(code.substr(9, 1)).toEqual("n")
	})

	it("matches the snapshot for the full alphabet", function () {
		const code = encodeCode39(Object.keys(Mapping).filter(it => it !== "*").sort().join(""))
		expect(code).toMatchSnapshot()
	})

	it("uppercases input strings", function () {
		const code = encodeCode39("a")
		expect(code).not.toEqual(expect.stringContaining(Mapping["-"]))
	})

	it("replaces out of range chars", function () {
		const code = encodeCode39("_")
		expect(code.substr(10,9)).toEqual(Mapping["-"])
	})

	it("replaces out of range chars with custom char", function () {
		const code = encodeCode39("_", " ")
		expect(code.substr(10,9)).toEqual(Mapping[" "])
	})

	describe("Mapping", function () {
		it("has 39 + 1 codes", function () {
			expect(Object.keys(Mapping).length).toEqual(40)
		})

		Object.keys(Mapping).map(char => {
			it(`matches the snapshot for '${char}'`, function () {
				expect(Mapping[char]).toMatchSnapshot()
			})

			it(`maps '${char}' to 5 bars`, function () {
				expect(Mapping[char].split("").filter(it => it === NARROW_BAR || it === WIDE_BAR)).toHaveLength(5)
			})

			it(`maps '${char}' to 4 spaces`, function () {
				expect(Mapping[char].split("").filter(it => it === NARROW_SPACE || it === WIDE_SPACE)).toHaveLength(4)
			})
		})
	})
})
