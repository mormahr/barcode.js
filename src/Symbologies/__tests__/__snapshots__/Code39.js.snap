// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`encodeCode39 Mapping matches the snapshot for ' ' 1`] = `
Array [
  "N",
  "w",
  "W",
  "n",
  "N",
  "n",
  "W",
  "n",
  "N",
]
`;

exports[`encodeCode39 Mapping matches the snapshot for '$' 1`] = `
Array [
  "N",
  "w",
  "N",
  "w",
  "N",
  "w",
  "N",
  "n",
  "N",
]
`;

exports[`encodeCode39 Mapping matches the snapshot for '%' 1`] = `
Array [
  "N",
  "n",
  "N",
  "w",
  "N",
  "w",
  "N",
  "w",
  "N",
]
`;

exports[`encodeCode39 Mapping matches the snapshot for '*' 1`] = `
Array [
  "N",
  "w",
  "N",
  "n",
  "W",
  "n",
  "W",
  "n",
  "N",
]
`;

exports[`encodeCode39 Mapping matches the snapshot for '+' 1`] = `
Array [
  "N",
  "w",
  "N",
  "n",
  "N",
  "w",
  "N",
  "w",
  "N",
]
`;

exports[`encodeCode39 Mapping matches the snapshot for '.' 1`] = `
Array [
  "W",
  "w",
  "N",
  "n",
  "N",
  "n",
  "W",
  "n",
  "N",
]
`;

exports[`encodeCode39 Mapping matches the snapshot for '/' 1`] = `
Array [
  "N",
  "w",
  "N",
  "w",
  "N",
  "n",
  "N",
  "w",
  "N",
]
`;

exports[`encodeCode39 Mapping matches the snapshot for '-' 1`] = `
Array [
  "N",
  "w",
  "N",
  "n",
  "N",
  "n",
  "W",
  "n",
  "W",
]
`;

exports[`encodeCode39 Mapping matches the snapshot for '0' 1`] = `
Array [
  "N",
  "n",
  "N",
  "w",
  "W",
  "n",
  "W",
  "n",
  "N",
]
`;

exports[`encodeCode39 Mapping matches the snapshot for '1' 1`] = `
Array [
  "W",
  "n",
  "N",
  "w",
  "N",
  "n",
  "N",
  "n",
  "W",
]
`;

exports[`encodeCode39 Mapping matches the snapshot for '2' 1`] = `
Array [
  "N",
  "n",
  "W",
  "w",
  "N",
  "n",
  "N",
  "n",
  "W",
]
`;

exports[`encodeCode39 Mapping matches the snapshot for '3' 1`] = `
Array [
  "W",
  "n",
  "W",
  "w",
  "N",
  "n",
  "N",
  "n",
  "N",
]
`;

exports[`encodeCode39 Mapping matches the snapshot for '4' 1`] = `
Array [
  "N",
  "n",
  "N",
  "w",
  "W",
  "n",
  "N",
  "n",
  "W",
]
`;

exports[`encodeCode39 Mapping matches the snapshot for '5' 1`] = `
Array [
  "W",
  "n",
  "N",
  "w",
  "W",
  "n",
  "N",
  "n",
  "N",
]
`;

exports[`encodeCode39 Mapping matches the snapshot for '6' 1`] = `
Array [
  "N",
  "n",
  "W",
  "w",
  "W",
  "n",
  "N",
  "n",
  "N",
]
`;

exports[`encodeCode39 Mapping matches the snapshot for '7' 1`] = `
Array [
  "N",
  "n",
  "N",
  "w",
  "N",
  "n",
  "W",
  "n",
  "W",
]
`;

exports[`encodeCode39 Mapping matches the snapshot for '8' 1`] = `
Array [
  "W",
  "n",
  "N",
  "w",
  "N",
  "n",
  "W",
  "n",
  "N",
]
`;

exports[`encodeCode39 Mapping matches the snapshot for '9' 1`] = `
Array [
  "N",
  "n",
  "W",
  "w",
  "N",
  "n",
  "W",
  "n",
  "N",
]
`;

exports[`encodeCode39 Mapping matches the snapshot for 'A' 1`] = `
Array [
  "W",
  "n",
  "N",
  "n",
  "N",
  "w",
  "N",
  "n",
  "W",
]
`;

exports[`encodeCode39 Mapping matches the snapshot for 'B' 1`] = `
Array [
  "N",
  "n",
  "W",
  "n",
  "N",
  "w",
  "N",
  "n",
  "W",
]
`;

exports[`encodeCode39 Mapping matches the snapshot for 'C' 1`] = `
Array [
  "W",
  "n",
  "W",
  "n",
  "N",
  "w",
  "N",
  "n",
  "N",
]
`;

exports[`encodeCode39 Mapping matches the snapshot for 'D' 1`] = `
Array [
  "N",
  "n",
  "N",
  "n",
  "W",
  "w",
  "N",
  "n",
  "W",
]
`;

exports[`encodeCode39 Mapping matches the snapshot for 'E' 1`] = `
Array [
  "W",
  "n",
  "N",
  "n",
  "W",
  "w",
  "N",
  "n",
  "N",
]
`;

exports[`encodeCode39 Mapping matches the snapshot for 'F' 1`] = `
Array [
  "N",
  "n",
  "W",
  "n",
  "W",
  "w",
  "N",
  "n",
  "N",
]
`;

exports[`encodeCode39 Mapping matches the snapshot for 'G' 1`] = `
Array [
  "N",
  "n",
  "N",
  "n",
  "N",
  "w",
  "W",
  "n",
  "W",
]
`;

exports[`encodeCode39 Mapping matches the snapshot for 'H' 1`] = `
Array [
  "W",
  "n",
  "N",
  "n",
  "N",
  "w",
  "W",
  "n",
  "N",
]
`;

exports[`encodeCode39 Mapping matches the snapshot for 'I' 1`] = `
Array [
  "N",
  "n",
  "W",
  "n",
  "N",
  "w",
  "W",
  "n",
  "N",
]
`;

exports[`encodeCode39 Mapping matches the snapshot for 'J' 1`] = `
Array [
  "N",
  "n",
  "N",
  "n",
  "W",
  "w",
  "W",
  "n",
  "N",
]
`;

exports[`encodeCode39 Mapping matches the snapshot for 'K' 1`] = `
Array [
  "W",
  "n",
  "N",
  "n",
  "N",
  "n",
  "N",
  "w",
  "W",
]
`;

exports[`encodeCode39 Mapping matches the snapshot for 'L' 1`] = `
Array [
  "N",
  "n",
  "W",
  "n",
  "N",
  "n",
  "N",
  "w",
  "W",
]
`;

exports[`encodeCode39 Mapping matches the snapshot for 'M' 1`] = `
Array [
  "W",
  "n",
  "W",
  "n",
  "N",
  "n",
  "N",
  "w",
  "N",
]
`;

exports[`encodeCode39 Mapping matches the snapshot for 'N' 1`] = `
Array [
  "N",
  "n",
  "N",
  "n",
  "W",
  "n",
  "N",
  "w",
  "W",
]
`;

exports[`encodeCode39 Mapping matches the snapshot for 'O' 1`] = `
Array [
  "W",
  "n",
  "N",
  "n",
  "W",
  "n",
  "N",
  "w",
  "N",
]
`;

exports[`encodeCode39 Mapping matches the snapshot for 'P' 1`] = `
Array [
  "N",
  "n",
  "W",
  "n",
  "W",
  "n",
  "N",
  "w",
  "N",
]
`;

exports[`encodeCode39 Mapping matches the snapshot for 'Q' 1`] = `
Array [
  "N",
  "n",
  "N",
  "n",
  "N",
  "n",
  "W",
  "w",
  "W",
]
`;

exports[`encodeCode39 Mapping matches the snapshot for 'R' 1`] = `
Array [
  "W",
  "n",
  "N",
  "n",
  "N",
  "n",
  "W",
  "w",
  "N",
]
`;

exports[`encodeCode39 Mapping matches the snapshot for 'S' 1`] = `
Array [
  "N",
  "n",
  "W",
  "n",
  "N",
  "n",
  "W",
  "w",
  "N",
]
`;

exports[`encodeCode39 Mapping matches the snapshot for 'T' 1`] = `
Array [
  "N",
  "n",
  "N",
  "n",
  "W",
  "n",
  "W",
  "w",
  "N",
]
`;

exports[`encodeCode39 Mapping matches the snapshot for 'U' 1`] = `
Array [
  "W",
  "w",
  "N",
  "n",
  "N",
  "n",
  "N",
  "n",
  "W",
]
`;

exports[`encodeCode39 Mapping matches the snapshot for 'V' 1`] = `
Array [
  "N",
  "w",
  "W",
  "n",
  "N",
  "n",
  "N",
  "n",
  "W",
]
`;

exports[`encodeCode39 Mapping matches the snapshot for 'W' 1`] = `
Array [
  "W",
  "w",
  "W",
  "n",
  "N",
  "n",
  "N",
  "n",
  "N",
]
`;

exports[`encodeCode39 Mapping matches the snapshot for 'X' 1`] = `
Array [
  "N",
  "w",
  "N",
  "n",
  "W",
  "n",
  "N",
  "n",
  "W",
]
`;

exports[`encodeCode39 Mapping matches the snapshot for 'Y' 1`] = `
Array [
  "W",
  "w",
  "N",
  "n",
  "W",
  "n",
  "N",
  "n",
  "N",
]
`;

exports[`encodeCode39 Mapping matches the snapshot for 'Z' 1`] = `
Array [
  "N",
  "w",
  "W",
  "n",
  "W",
  "n",
  "N",
  "n",
  "N",
]
`;

exports[`encodeCode39 matches the snapshot for the full alphabet 1`] = `
Array [
  "N",
  "w",
  "N",
  "n",
  "W",
  "n",
  "W",
  "n",
  "N",
  "n",
  "N",
  "w",
  "W",
  "n",
  "N",
  "n",
  "W",
  "n",
  "N",
  "n",
  "N",
  "w",
  "N",
  "w",
  "N",
  "w",
  "N",
  "n",
  "N",
  "n",
  "N",
  "n",
  "N",
  "w",
  "N",
  "w",
  "N",
  "w",
  "N",
  "n",
  "N",
  "w",
  "N",
  "n",
  "N",
  "w",
  "N",
  "w",
  "N",
  "n",
  "N",
  "w",
  "N",
  "n",
  "N",
  "n",
  "W",
  "n",
  "W",
  "n",
  "W",
  "w",
  "N",
  "n",
  "N",
  "n",
  "W",
  "n",
  "N",
  "n",
  "N",
  "w",
  "N",
  "w",
  "N",
  "n",
  "N",
  "w",
  "N",
  "n",
  "N",
  "n",
  "N",
  "w",
  "W",
  "n",
  "W",
  "n",
  "N",
  "n",
  "W",
  "n",
  "N",
  "w",
  "N",
  "n",
  "N",
  "n",
  "W",
  "n",
  "N",
  "n",
  "W",
  "w",
  "N",
  "n",
  "N",
  "n",
  "W",
  "n",
  "W",
  "n",
  "W",
  "w",
  "N",
  "n",
  "N",
  "n",
  "N",
  "n",
  "N",
  "n",
  "N",
  "w",
  "W",
  "n",
  "N",
  "n",
  "W",
  "n",
  "W",
  "n",
  "N",
  "w",
  "W",
  "n",
  "N",
  "n",
  "N",
  "n",
  "N",
  "n",
  "W",
  "w",
  "W",
  "n",
  "N",
  "n",
  "N",
  "n",
  "N",
  "n",
  "N",
  "w",
  "N",
  "n",
  "W",
  "n",
  "W",
  "n",
  "W",
  "n",
  "N",
  "w",
  "N",
  "n",
  "W",
  "n",
  "N",
  "n",
  "N",
  "n",
  "W",
  "w",
  "N",
  "n",
  "W",
  "n",
  "N",
  "n",
  "W",
  "n",
  "N",
  "n",
  "N",
  "w",
  "N",
  "n",
  "W",
  "n",
  "N",
  "n",
  "W",
  "n",
  "N",
  "w",
  "N",
  "n",
  "W",
  "n",
  "W",
  "n",
  "W",
  "n",
  "N",
  "w",
  "N",
  "n",
  "N",
  "n",
  "N",
  "n",
  "N",
  "n",
  "W",
  "w",
  "N",
  "n",
  "W",
  "n",
  "W",
  "n",
  "N",
  "n",
  "W",
  "w",
  "N",
  "n",
  "N",
  "n",
  "N",
  "n",
  "W",
  "n",
  "W",
  "w",
  "N",
  "n",
  "N",
  "n",
  "N",
  "n",
  "N",
  "n",
  "N",
  "w",
  "W",
  "n",
  "W",
  "n",
  "W",
  "n",
  "N",
  "n",
  "N",
  "w",
  "W",
  "n",
  "N",
  "n",
  "N",
  "n",
  "W",
  "n",
  "N",
  "w",
  "W",
  "n",
  "N",
  "n",
  "N",
  "n",
  "N",
  "n",
  "W",
  "w",
  "W",
  "n",
  "N",
  "n",
  "W",
  "n",
  "N",
  "n",
  "N",
  "n",
  "N",
  "w",
  "W",
  "n",
  "N",
  "n",
  "W",
  "n",
  "N",
  "n",
  "N",
  "w",
  "W",
  "n",
  "W",
  "n",
  "W",
  "n",
  "N",
  "n",
  "N",
  "w",
  "N",
  "n",
  "N",
  "n",
  "N",
  "n",
  "W",
  "n",
  "N",
  "w",
  "W",
  "n",
  "W",
  "n",
  "N",
  "n",
  "W",
  "n",
  "N",
  "w",
  "N",
  "n",
  "N",
  "n",
  "W",
  "n",
  "W",
  "n",
  "N",
  "w",
  "N",
  "n",
  "N",
  "n",
  "N",
  "n",
  "N",
  "n",
  "W",
  "w",
  "W",
  "n",
  "W",
  "n",
  "N",
  "n",
  "N",
  "n",
  "W",
  "w",
  "N",
  "n",
  "N",
  "n",
  "W",
  "n",
  "N",
  "n",
  "W",
  "w",
  "N",
  "n",
  "N",
  "n",
  "N",
  "n",
  "W",
  "n",
  "W",
  "w",
  "N",
  "n",
  "W",
  "w",
  "N",
  "n",
  "N",
  "n",
  "N",
  "n",
  "W",
  "n",
  "N",
  "w",
  "W",
  "n",
  "N",
  "n",
  "N",
  "n",
  "W",
  "n",
  "W",
  "w",
  "W",
  "n",
  "N",
  "n",
  "N",
  "n",
  "N",
  "n",
  "N",
  "w",
  "N",
  "n",
  "W",
  "n",
  "N",
  "n",
  "W",
  "n",
  "W",
  "w",
  "N",
  "n",
  "W",
  "n",
  "N",
  "n",
  "N",
  "n",
  "N",
  "w",
  "W",
  "n",
  "W",
  "n",
  "N",
  "n",
  "N",
  "n",
  "N",
  "w",
  "N",
  "n",
  "W",
  "n",
  "W",
  "n",
  "N",
]
`;
